cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,320 @@
+   * {@link ActionAcceptance#BROKER_REJECT} if the action is rejected due to violating brokerSet awareness in the destination
+   * broker after moving source replica to destination broker, {@link ActionAcceptance#REPLICA_REJECT} otherwise.
+   */
+  @Override
+  public ActionAcceptance actionAcceptance(BalancingAction action, ClusterModel clusterModel) {
+    switch (action.balancingAction()) {
+      case LEADERSHIP_MOVEMENT:
+        return ACCEPT;
+      case INTER_BROKER_REPLICA_MOVEMENT:
+      case INTER_BROKER_REPLICA_SWAP:
+        if (doesReplicaMoveViolateActionAcceptance(clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition()),
+                                                   clusterModel.broker(action.destinationBrokerId()))) {
+          return BROKER_REJECT;
+        }
+        // SWAP Actions are performed by distribution goals to swap distinct partitions replicas to swap load in or out
+        // We would need to check both ways in swap cases that original brokerSet of both replicas matches the brokerSet of destination broker
+        if (action.balancingAction() == ActionType.INTER_BROKER_REPLICA_SWAP && doesReplicaMoveViolateActionAcceptance(
+            clusterModel.broker(action.destinationBrokerId()).replica(action.destinationTopicPartition()),
+            clusterModel.broker(action.sourceBrokerId()))) {
+          return REPLICA_REJECT;
+        }
+        return ACCEPT;
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  protected boolean doesReplicaMoveViolateActionAcceptance(Replica sourceReplica, Broker destinationBroker) {
+    // Destination broker cannot be in any other brokerSet.
+    // A replica's current broker can be different from the original broker after re-balance by other goals
+    // Since the replica's original broker is the correct indicator of what brokerSet the replica belongs to, we use
+    // sourceReplica.originalBroker().
+    try {
+      String originalSourceBrokerSetId = _brokerSetResolutionHelper.brokerSetId(sourceReplica.originalBroker());
+      String destinationBrokerSetId = _brokerSetResolutionHelper.brokerSetId(destinationBroker);
+
+      return !originalSourceBrokerSetId.equals(destinationBrokerSetId);
+    } catch (BrokerSetResolutionException e) {
+      LOG.error(String.format("[%s] ", name()), e);
+      return true;
+    }
+  }
+
+  @Override
+  public ClusterModelStatsComparator clusterModelStatsComparator() {
+    // Stats are irrelevant to a hard goal. The optimization would already fail if the goal requirements are not met.
+    return new GoalUtils.HardGoalStatsComparator();
+  }
+
+  @Override
+  public ModelCompletenessRequirements clusterModelCompletenessRequirements() {
+    // BrokerSet-awareness does not need load information at all
+    return new ModelCompletenessRequirements(MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING, 0.0, true);
```

Comment:
I'm wondering if we need MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING or just 0 for this goal. We probably don't need any valid windows for this goal?

Reply:
That's correct, we need 0 just like RackAwareGoal. This Goal requires information about BrokerSet to brokerId mapping, Topic to BrokerSet mapping and meta data client. 
MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING is needed for goals that use broker and partition metrics.

Reply:
Yeah exactly, but in that case shouldn't we replace MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING with 0 here?

Reply:
That is what we are doing here. We are saying the MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING is 0 for this Goal to consider model complete. Could you please explain what do you mean by replacing the value by 0 ?

Reply:
Ah ok. So I checked the code, and the value of MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING is 1 (defined in GoalUtils.java). I was saying it shouldn't be 1 because we don't need any valid windows here. Instead, it should 0. So the code should look like:

`return new ModelCompletenessRequirements(0, 0.0, true);`

Reply:
That won't work - https://github.com/linkedin/cruise-control/blob/4e5927b48bf2581ab76acbbecbf42b355b871b65/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/ModelCompletenessRequirements.java#L57

Reply:
Ah got it! Thanks for the explanation. I'll mark it as resolved.

Reply:
I have updated my earlier comment and corrected - it is actually **minValidPartitionsRatio** that we need 0 and not **MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING**.